pkg/gateway/model_build_targetgroup.go

```diff
@@ -247,7 +256,7 @@ func (t *backendRefTargetGroupModelBuildTask) buildTargetGroup(ctx context.Conte
 	}
 	t.log.Debugf("Added target group for backendRef %s to the stack %s", t.backendRef.Name(), stackTG.ID())
 
-	stackTG.IsDeleted = !t.route.DeletionTimestamp().IsZero()
+	stackTG.IsDeleted = !t.route.DeletionTimestamp().IsZero() // should always be false
```

Comment:
QQ: now, which part of controller code is responsible to do "delete the k8s Service trigger deleting Unused Target Groups"? I think should be  `targetGroupSynthesizer.SynthesizeUnusedDelete()` ?

Seems to understand changes in one function also need to have heavy knowledge on how other functions works... a little bit hard to trace code..

Reply:
Yes, unfortunately at the moment you kind of need to know what's happening elsewhere to know if what's being done locally is OK. The unit tests help explain some of that and enforce the invariants. Part of the issue with keeping this logic entirely separate was we were doing a lot of redundant/throw-away work. For example, we would create models for listeners that were going to be deleted then do nothing with them. I think this is a good example of optimisation being at odds with properties like decoupling/modularization. Having said that, I think we can look at making this better. My intuition is that as part of #463 we'll be able to have better performance AND readability.

To answer the specific question, any unused target groups or target groups which reference invalid services will be deleted as part of the ```SynthesizeUnusedDelete()```. We can and should run this cleanup on a schedule rather than only when a change is made.